File: cpp/src/0019.cpp
/*
[19. Remove Nth Node From End of List (Medium)](https://leetcode.com/problems/remove-nth-node-from-end-of-list/)

@topic: Linked List
@topic: Two Pointers

Constraints:
  - The number of nodes in the list is `sz`.
  - `1 <= sz <= 30`
  - `0 <= Node.val <= 100`
  - `1 <= n <= sz`
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        auto fake = ListNode(0, head);
        auto slow = &fake;
        auto fast = head;
        while (n--) fast = fast->next;
        while (fast) {
            slow = slow->next;
            fast = fast->next;
        }
        auto next = slow->next;
        slow->next = next->next;
        delete next;
        return fake.next;
    }

    /* ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (dfs(head, n) == n) {
            auto next = head->next;
            head->next = next;
            delete head;
            return next;
        } else {
            return head;
        }
    }

   private:
    int dfs(ListNode* head, int n) {
        if (!head) return 0;
        int x = dfs(head->next, n);
        if (x == n) {
            auto next = head->next;
            head->next = next->next;
            delete next;
        }
        return x + 1;
    } */
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, int, ListNode*>> CASES = {
        {of({1, 2, 3, 4, 5}), 2, of({1, 2, 3, 5})},
        {of({1}), 1, of({})},
        {of({1, 2}), 1, of({1})},
    };

    for (auto& [head, n, excepted] : CASES) {
        assert(eq(o.removeNthFromEnd(head, n), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.removeNthFromEnd(of({1, 2, 3, 4, 5}), 2);
        o.removeNthFromEnd(of({1}), 1);
        o.removeNthFromEnd(of({1, 2}), 1);
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0019.py
from typing import Optional
from structs.ListNode import ListNode, create


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fake = ListNode(0, head)
        slow = fake
        fast = head
        for _ in range(n):
            fast = fast.next
        while fast:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return fake.next

    # def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
    #     return head.next if self.__dfs(head, n) == n else head

    # def __dfs(self, head: Optional[ListNode], n: int):
    #     if not head:
    #         return 0
    #     x = self.__dfs(head.next, n)
    #     if x == n:
    #         head.next = head.next.next
    #     return x + 1


if __name__ == '__main__':
    import unittest

    CASES = (
        (create(1, 2, 3, 4, 5), 2, create(1, 2, 3, 5)),
        (create(1), 1, None),
        (create(1, 2), 1, create(1)),
    )

    class Test(unittest.TestCase):

        def test(self):
            removeNthFromEnd = Solution().removeNthFromEnd
            for head, n, excepted in CASES:
                actual = removeNthFromEnd(head, n)
                self.assertEqual(actual, excepted)

    unittest.main()
